tests/test_sync.py

```diff
@@ -137,17 +157,123 @@ def test_copy_part_gs_to_s3(self):
+        check_file_revdeps()
+
+        blob_key = compose_blob_key(file_data)
+        blob_blob = self.s3_bucket.Object(blob_key)
+        blob_blob.put(Body=b"sync_test")
+
+        @eventually(timeout=8, interval=1, errors={Exception})
+        def check_blob_revdeps():
+            self.assertTrue(sync.dependencies_exist(Replica.aws, Replica.aws, collection_key))
+            self.assertTrue(sync.dependencies_exist(Replica.aws, Replica.aws, bundle_key))
+            self.assertTrue(sync.dependencies_exist(Replica.aws, Replica.aws, file_key))
+        check_blob_revdeps()
+
+# TODO: (akislyuk) integration test of SQS fault injection, SFN fault injection
+
+@testmode.integration
+class TestSyncDaemon(unittest.TestCase, TestSync):
+    def setUp(self):
+        dss.Config.set_config(dss.BucketConfig.TEST)
+        self.gs_bucket_name, self.s3_bucket_name = dss.Config.get_gs_bucket(), dss.Config.get_s3_bucket()
+        self.logger = logging.getLogger(__name__)
+        self.gs = Config.get_native_handle(Replica.gcp)
+        self.gs_bucket = self.gs.bucket(self.gs_bucket_name)
+        self.s3 = boto3.resource("s3")
+        self.s3_bucket = self.s3.Bucket(self.s3_bucket_name)
+
+    def test_sync_small_blob(self):
+        "Tests oneshot blob syncing."
+        self._sync_blob()
+
+    @unittest.skipUnless(json.loads(os.environ.get("DSS_TEST_LARGE_FILE_SYNC", "false")),
+                         "Skipping large file sync test. Set DSS_TEST_LARGE_FILE_SYNC=1 to run it")
+    def test_sync_large_blob(self):
+        "Tests multipart blob syncing."
+        self._sync_blob(payload_size=max(sync.part_size.values()) + 1)
+
+    def _sync_blob(self, payload_size=8):
+        self.cleanup_sync_test_objects()
+        payload = self.get_payload(payload_size)
+        test_metadata = {"metadata-sync-test": str(uuid.uuid4())}
+        test_key = "{}/s3-to-gcs/{}".format(self.test_blob_prefix, uuid.uuid4())
+        src_blob = self.s3_bucket.Object(test_key)
+        gs_dest_blob = self.gs_bucket.blob(test_key)
+        src_blob.put(Body=payload, Metadata=test_metadata)
+
+        @eventually(timeout=60, interval=1, errors={Exception})
+        def check_gs_dest():
+            self.assertEqual(gs_dest_blob.download_as_string(), payload)
+        check_gs_dest()
+
+        test_key = "{}/gcs-to-s3/{}".format(self.test_blob_prefix, uuid.uuid4())
+        src_blob = self.gs_bucket.blob(test_key)
+        dest_blob = self.s3_bucket.Object(test_key)
+        src_blob.metadata = test_metadata
+        src_blob.upload_from_string(payload)
+
+        @eventually(timeout=60, interval=1, errors={Exception})
+        def check_s3_dest(dest_blob):
+            self.assertEqual(dest_blob.get()["Body"].read(), payload)
+            self.assertEqual(dest_blob.metadata, test_metadata)
+        check_s3_dest(dest_blob)
+
+        # GS metadata seems to take a while to propagate, so we wait until the above test completes to read it back
+        gs_dest_blob.reload()
```

Comment:
same thing as with `TestSyncUtils.test_sync_blob`

Reply:
In both these cases, I have moved the metadata check up to immediately follow the previous test clauses and not use the other operations as a delay. The tests passed as a result. So at this point I consider this workaround unnecessary.

I think the practice of using event sequencing this way for testing is not bad, as long as it's properly documented (as it is here) and is (exceedingly) unlikely to cause race conditions (as it is here). For me it certainly works better than splitting tests into multiple files (increasing the code size, splitting topically related code, and making it harder to understand).